### PR TITLE
Update DWD optimizers

### LINKED ISSUE #1334: Composer prevents loss from converging when using DecoupledAdamW
** Environment **
- OS: Ubuntu 22.04 LTS
- GPU:
```
  *-display                 
       description: VGA compatible controller
       product: GP102 [GeForce GTX 1080 Ti]
       vendor: NVIDIA Corporation
       physical id: 0
       bus info: pci@0000:01:00.0
       version: a1
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress vga_controller bus_master cap_list rom
       configuration: driver=nvidia latency=0
       resources: irq:58 memory:fa000000-faffffff memory:c0000000-cfffffff memory:d0000000-d1ffffff ioport:e000(size=128) memory:c0000-dffff
  *-graphics
       product: EFI VGA
       physical id: 2
       logical name: /dev/fb0
       capabilities: fb
       configuration: depth=32 resolution=1024,768
```
- Cuda: 11.7
- Composer: 0.8.2

I'm trying to implement a paper in Pytorch, using composer for my training loop.
See here: https://github.com/vedantroy/improved-ddpm-pytorch/commit/24909c064953bf54bf3a50c64923fdfef714f3cc

Composer is breaking stuff in at least 2 different ways:
1. If I replace this line with composer's `DecoupledAdamW` then the loss goes from 1.0 => 0.6 but not much further down
   https://github.com/vedantroy/improved-ddpm-pytorch/blob/24909c064953bf54bf3a50c64923fdfef714f3cc/openai_iddpm.py#L103
2. ~If I use the `OpenAIIDDPM` model instead of the `manual_train` function, the loss does not go down from around 1.0 at all
  https://github.com/vedantroy/improved-ddpm-pytorch/blob/24909c064953bf54bf3a50c64923fdfef714f3cc/openai_iddpm.py#L72~
 - I think only the 1st thing was a bug b/c once I switched to torch's `AdamW` and removed the warmup, the learning rate progressed as normal

I've uploaded the data here, so you can reproduce it: https://drive.google.com/drive/folders/1zK-3SkGOETGFajIBSXbEPqPSRdwqajUq?usp=sharing

It's just the CelebHQ dataset in 64x64 resolution, stored as parquet. Any image dataset should work and you can probably swap out my torch data based dataloader for something more standard.

### PR DESCRIPTION
**Pull request by abhi-mosaic**
This PR fixes #1334 and does the following:

* fix bug when `lr=0`
* Fix some typings
* Update `DecoupledAdamW` default values for `weight_decay` and `betas` to be safer

@ravi-mosaicml I'm wondering if I should also go through all the docstrings and clean them up? I see some inconsistent formatting like `weight_decay (float, optional)` and wonder if it should be `weight_decay (Optional[float])`. Let me know what you suggest!

### CODE REVIEW

**Comment by hanlint** [comment]
I prefer `float, optional` as more readable than `Optional[float]`

**Comment by ravi-mosaicml** [comment]
> I prefer `float, optional` as more readable than `Optional[float]`

Agreed

**Review by ravi-mosaicml — COMMENTED** [review]
Code changes lgtm; see comments. Will let someone from research sign off on it.

File: composer/optim/decoupled_weight_decay.py [hunk 1]
```diff
@@ -160,12 +166,28 @@ class DecoupledAdamW(AdamW):
...
-                                 Default: ``(0.9, 0.999)``.
+                                 Default: ``(0.9, 0.95)``.
         eps (float, optional): Term added to the denominator to improve numerical stability. Default: ``1e-8``.
-        weight_decay (float, optional): Decoupled weight decay factor. Default: ``1e-2``.
+        weight_decay (float, optional): Decoupled weight decay factor. Default: ``1e-5``.
         amsgrad (bool, optional): Enables the amsgrad variant of Adam. Default: ``False``.
     """
 
+    def __init__(self,
+                 params: List[torch.Tensor],
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/optim/decoupled_weight_decay.py:176]: Interesting... I'm reading through `Optimizer`'s init and it doesn't have types :( but it seems like `params` should accept an iterable of `Tensors` or `dicts`, I'm not sure if `Parameter` needs to be specified as it is a subclass of `Tensor`?

**Comment by abhi-mosaic** [comment]
> > I prefer `float, optional` as more readable than `Optional[float]`
> 
> Agreed

Random question, if the docstring says `(float, optional)`, that seems to imply you can pass `None`, but that is actually not allowed, and the type is `float` not `Optional[float]`. Is this just a quirk or should the docstring really be `(float)`, and say that there is a default?

File: composer/optim/decoupled_weight_decay.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/optim/decoupled_weight_decay.py]: Updated to use type `torch.optimizer._params_t` which is basically a shortcut for:
```python
_params_t = Union[Iterable[Tensor], Iterable[dict]]
```

**Comment by hanlint** [comment]
The `optional` in our docstrings means that it has a default (e.g. similar to pytorch: https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html), and is not a required argument. 

If something has the type signature `Optional[X]`, then the docstrings should read (X or None).

**Review by moinnadeem — APPROVED** [review]
LGTM -- how would we also feel about a `log.warning(...)` when the WD is pretty high, and it might indicate that the user hasn't decoupled their values? I'm thinking about usability where a user just copies and pastes values from AdamW in PyTorch -- how can we catch that happening and help them avoid it?

**Comment by abhi-mosaic** [comment]
> LGTM -- how would we also feel about a `log.warning(...)` when the WD is pretty high, and it might indicate that the user hasn't decoupled their values? I'm thinking about usability where a user just copies and pastes values from AdamW in PyTorch -- how can we catch that happening and help them avoid it?

That's a great point, I will add that before merging.
